### PR TITLE
git module

### PR DESCRIPTION
**Pull request by samstav**

### CODE REVIEW

File: simpl/git.py [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+
+
+def resolve_git_reference(gitrepo, ref, remote='origin'):
+    """Try to find a revision (commit hash) that corresponds to 'ref'.
+
+    Even if 'ref' is not fetched, use a combination of fetch and/or
+    ls-remote to discover it and check it out.
+
+    Note:
+        stole these ideas from opscode chef, chef/provider/git.rb
```
Comment (samstav) [inline simpl/git.py]: they changed their name to Chef

File: simpl/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    """Raised when an error occurs while trying a git command."""
+
+    def __init__(self, returncode, cmd, output=None, oserror=None):
+        super(SimplGitCommandError, self).__init__()
+        self.returncode = returncode
+        self.cmd = cmd
+        self.output = output
+        self.oserror = oserror
+        super(SimplGitCommandError, self).__init__()
```
Comment (ziadsawalha) [inline simpl/exceptions.py]: Duplicate from L64?

**Comment by larsbutler** [comment]
Hey @samstav, test coverage looks pretty good:

```
Name                Stmts   Miss  Cover   Missing
-------------------------------------------------
simpl                   8      0   100%   
simpl.config          324     43    87%   166-167, 245, 250-251, 316-319, 325, 361-363, 385, 414, 418, 422, 426, 433, 476, 499-500, 543-547, 562, 568, 596-597, 604, 606-609, 620-621, 626, 732, 742, 765, 770-787
simpl.exceptions       33      8    76%   40, 71, 76-81, 104
simpl.git             283     14    95%   135, 194-195, 251, 303, 401, 446, 448, 511, 657-661
simpl.utils             0      0   100%   
simpl.utils.shell      25      1    96%   66
-------------------------------------------------
TOTAL                 673     66    90%   
----------------------------------------------------------------------
```

Unless @ziadsawalha has any objections (since he started the review), I would say, let's merge it.

**Comment by samstav** [comment]
FYI, this caused some confusion:

http://stackoverflow.com/a/20608181/1547030

Tests were failing on travis because I was `git fetch`ing with `--tags`. On git versions < 1.9 if you use `--tags` in your fetch, it means _only_ fetch tags.

**Pull request merged**